### DISCUSSION TITLE
Many calls to Plotter leads to segmentation fault: Maximum number of clients reached?

### DISCUSSION THREAD

**Discussion by leo2r** [pyvista/pyvista#3203] [Q&A] [answered]
Hi all,

I've read up on other issues about this, and it seemed like the problem may have been fixed in newer versions of VTK and PyVista, but I am still having problems when calling Plotter many times in a for-loop. Below is my code where I call the plotter:

```python
pl = pv.Plotter(off_screen=True) # Initialised at the top of the Class

obs = self.create_obs_dict() # In the step() function, called many times at runtime

def create_obs_dict(self):
        obs = {}
        obs['img1'] = self.get_image(cpos=[0, 0, 250], croll=0)
        obs['img2'] = self.get_image(cpos=[200, 0, 0], croll=-90)
        obs['img3'] = self.get_image(cpos=[-200, 0, 0], croll=90)
        return obs
        
def get_image(self, cpos=[0, 0, 0], croll=0):
        self.pl.clear()
        self.pl.deep_clean()
        self.pl.add_mesh(self.bone_mesh, color="lightblue", opacity=0.3) #style='wireframe',
        self.pl.add_mesh(self.screw1_mesh, color=True, opacity=1)
        self.pl.add_mesh(self.screw2_mesh, color=True, opacity=1)
        self.pl.camera.position = cpos
        self.pl.camera.focal_point = (0,0,0)
        self.pl.camera.roll = croll
        img = self.pl.screenshot(None, return_img=True, window_size=(self.WINDOW_SIZE))
        img = img[:, :, 0]
        return np.expand_dims(img, axis=2) # Correct format from (WIDTH, HEIGHT) to (WIDTH, HEIGHT, 1)
```

The method ```create_obs_dict()``` gets called once every ```step``` in my reinforcement learning program. It seems like the problem may be due to the plotter not closing correctly? What can I do? Whether I run the code on my Mac or on a Ubuntu VM, I get the *segmentation fault* or *abort (core dumped)*. I have ```pyvista == 0.36.1``` and ```vtk == 9.1.0```. Any help would be greatly appreciated, this is quite an urgent error I must deal with! 🙏🏽

**Comment by akaszynski** [comment, marked as answer]
Can you please check out the [refactor/render_window](https://github.com/pyvista/pyvista/tree/refactor/render_window) branch and see if that solves your issue? There are several objects that are not being collected and that should clean it up.

```
git checkout https://github.com/pyvista/pyvista.git
cd pyvista
git checkout refactor/render_window
pip install -e .
```

**Reply by leo2r** [reply]
Thanks, will try this now!

**Reply by leo2r** [reply]
I'm now getting an error:

```
Traceback (most recent call last):
  File "train.py", line 7, in <module>
    import pyvista as pv
  File "/Users/leoderiedmatten/pyvista/pyvista/__init__.py", line 12, in <module>
    from pyvista.plotting import *
  File "/Users/leoderiedmatten/pyvista/pyvista/plotting/__init__.py", line 4, in <module>
    from .charts import Chart2D, ChartMPL, ChartBox, ChartPie
  File "/Users/leoderiedmatten/pyvista/pyvista/plotting/charts.py", line 13, in <module>
    from pyvista import _vtk
  File "/Users/leoderiedmatten/pyvista/pyvista/_vtk.py", line 266, in <module>
    from vtkmodules.vtkFiltersParallelDIY2 import vtkRedistributeDataSetFilter
ModuleNotFoundError: No module named 'vtkmodules.vtkFiltersParallelDIY2'
```

Do I need to install a specific version of ```vtk``` to use this ```refactor/render_window``` branch? Cheers

**Reply by akaszynski** [reply]
We just addressed this issue in a PR. Please pull again, I merged with main.

**Reply by leo2r** [reply]
Okay I'll pull again, but should I stay on the refactor/render_window or switch back to main? Slightly confused sorry 😅

**Reply by akaszynski** [reply]
Correct, stay on `refactor/render_window`.

**Reply by leo2r** [reply]
Unfortunately getting another error now:

```
Traceback (most recent call last):
  File "train.py", line 82, in <module>
    model.learn(total_timesteps=2000000, callback=callback, tb_log_name = exp_name)
  File "/Users/leoderiedmatten/miniconda3/envs/t21-pyvista/lib/python3.7/site-packages/stable_baselines3/ppo/ppo.py", line 308, in learn
    reset_num_timesteps=reset_num_timesteps,
  File "/Users/leoderiedmatten/miniconda3/envs/t21-pyvista/lib/python3.7/site-packages/stable_baselines3/common/on_policy_algorithm.py", line 243, in learn
    total_timesteps, eval_env, callback, eval_freq, n_eval_episodes, eval_log_path, reset_num_timesteps, tb_log_name
  File "/Users/leoderiedmatten/miniconda3/envs/t21-pyvista/lib/python3.7/site-packages/stable_baselines3/common/base_class.py", line 429, in _setup_learn
    self._last_obs = self.env.reset()  # pytype: disable=annotation-type-mismatch
  File "/Users/leoderiedmatten/miniconda3/envs/t21-pyvista/lib/python3.7/site-packages/stable_baselines3/common/vec_env/vec_transpose.py", line 110, in reset
    return self.transpose_observations(self.venv.reset())
  File "/Users/leoderiedmatten/miniconda3/envs/t21-pyvista/lib/python3.7/site-packages/stable_baselines3/common/vec_env/dummy_vec_env.py", line 61, in reset
    obs = self.envs[env_idx].reset()
  File "/Users/leoderiedmatten/miniconda3/envs/t21-pyvista/lib/python3.7/site-packages/stable_baselines3/common/monitor.py", line 79, in reset
    return self.env.reset(**kwargs)
  File "/Users/leoderiedmatten/miniconda3/envs/t21-pyvista/lib/python3.7/site-packages/gym/envs/custom/pedscrewplan.py", line 189, in reset
    obs = self.create_obs_dict()
  File "/Users/leoderiedmatten/miniconda3/envs/t21-pyvista/lib/python3.7/site-packages/gym/envs/custom/pedscrewplan.py", line 153, in create_obs_dict
    obs['img1'] = self.get_image(cpos=[0, 0, 250], croll=0)
  File "/Users/leoderiedmatten/miniconda3/envs/t21-pyvista/lib/python3.7/site-packages/gym/envs/custom/pedscrewplan.py", line 164, in get_image
    self.pl.camera.position = cpos
  File "/Users/leoderiedmatten/pyvista/pyvista/plotting/plotting.py", line 1232, in camera
    self.camera_position = self.get_default_cam_pos()
  File "/Users/leoderiedmatten/pyvista/pyvista/plotting/plotting.py", line 1358, in camera_position
    self.renderer.camera_position = camera_location
  File "/Users/leoderiedmatten/pyvista/pyvista/plotting/renderer.py", line 311, in camera_position
    self.camera.position = scale_point(self.camera, camera_location[0], invert=False)
  File "/Users/leoderiedmatten/pyvista/pyvista/plotting/renderer.py", line 114, in scale_point
    mtx = camera.GetModelTransformMatrix()
AttributeError: 'NoneType' object has no attribute 'GetModelTransformMatrix'
```

If I comment out the ```pl.deep_clean()``` line I get the segmentation error instead:
```
Segmentation fault: 11
```

**Reply by akaszynski** [reply]
Interesting. Ok, I'll need a minimum reproducible example to fix the PR. I've got a Mac so I can at least test it out on the same platform.

If none of this works, consider trying (on `0.36.1`):
```py
import pyvista
pyvista.close_all()
```
However, I'd like this to work without having to resort to `deep_clean` and `close_all`.

**Reply by leo2r** [reply]
Okay, I'll clean up my code to create a minimum reproducible example! Will get back to you ASAP, thanks 🙏🏽

**Reply by leo2r** [reply]
So by trying to get rid of irrelevant code to create a minimal reproducible example, the segmentation fault has disappeared 😂 I unfortunately don't have time to continue right now, but I'll be back at it tomorrow and will get back in touch here as soon as I've figured out what's going on! Thanks already for your help so far :)

**Reply by leo2r** [reply]
Hi @akaszynski. I'm having trouble creating a minimal reproducible example as it seems the segmentation fault error sometimes happens, but other times the program just gets stuck somewhere and runs without any output... I'm trying to figure out what is going on but may take a little while...

**Reply by leo2r** [reply]
@akaszynski What about using Python's ```with``` function, eg:

```python
def get_images(self, cpos=[0, 0, 0], croll=0):
        with pv.Plotter(off_screen=True) as pl:
            pl.add_mesh(self.bone_mesh, color="lightblue", opacity=0.3) #style='wireframe',
            pl.add_mesh(self.screw1_mesh, color=True, opacity=1)
            pl.add_mesh(self.screw2_mesh, color=True, opacity=1)
            pl.camera.position = cpos
            pl.camera.focal_point = (0,0,0)
            pl.camera.roll = croll
            img = pl.screenshot(None, return_img=True, window_size=(self.WINDOW_SIZE))
            img = img[:, :, 0]
            return np.expand_dims(img, axis=2) # Correct format from (WIDTH, HEIGHT) to (WIDTH, HEIGHT, 1)
```

That should properly clean up everything, right?

**Reply by leo2r** [reply]
> @akaszynski What about using Python's `with` function, eg:
> 
> ```python
> def get_images(self, cpos=[0, 0, 0], croll=0):
>         with pv.Plotter(off_screen=True) as pl:
>             pl.add_mesh(self.bone_mesh, color="lightblue", opacity=0.3) #style='wireframe',
>             pl.add_mesh(self.screw1_mesh, color=True, opacity=1)
>             pl.add_mesh(self.screw2_mesh, color=True, opacity=1)
>             pl.camera.position = cpos
>             pl.camera.focal_point = (0,0,0)
>             pl.camera.roll = croll
>             img = pl.screenshot(None, return_img=True, window_size=(self.WINDOW_SIZE))
>             img = img[:, :, 0]
>             return np.expand_dims(img, axis=2) # Correct format from (WIDTH, HEIGHT) to (WIDTH, HEIGHT, 1)
> ```
> 
> That should properly clean up everything, right?

Actually maybe not since I'd need to implement the ```__exit__``` function which would require ```close()``` or ```deep_clean()``` etc...

**Reply by akaszynski** [reply]
I'm seeing a potential regression on my end, so your segfault might not be a fluke. See https://github.com/pyvista/pyvista/pull/3214

For the time being, can you try using `pl.deep_clean()` and `pyvista.close_all` on the latest release?

**Reply by leo2r** [reply]
Thanks for the suggestion, I'll try that. Just a few questions:

1. Should I create the ```Plotter()``` object as a class attribute and call ```pl.clear()``` and ```pl.deep_clean()``` in ```get_images()```? Or should I create a new instance of ```Plotter()``` at the start of ```get_images()```?
2. Where does ```pl.deep_clean()``` go? At the end of ```get_images()``` before the return statement?
3. Where do I call ```pv.close_all()```? Also at the end of ```get_images()``` before the return statement?

Thank you for your help!

**Reply by leo2r** [reply]
@akaszynski Could you check out the questions above please? :) Thanks!

**Reply by akaszynski** [reply]
- Create a new instance of `Plotter()` at the start of `get_images()`
- `pl.deep_clean()` and `pv.close_all()` right at the end of `get_images()`
- Keep all the references to your `pl` within `get_images` so that when the method returns `pl` goes out of scope.

The [refactor/render_window](https://github.com/pyvista/pyvista/tree/refactor/render_window) branch should fix this once I figure out where this transient segfault is coming from.

**Reply by leo2r** [reply]
Thanks so much, implementing this now!

**Reply by leo2r** [reply]
Hey, I'm getting a ```PyvistaDeprecationWarning: `deep_clean` has been deprecated, Use `Plotter.clear()``` warning, should I switch to ```clear()``` then, does it do the same thing?

I've tried with ```deep_clean()``` and ```clear()```, followed by ```pv.close_all()```. I sometimes get the segmentation fault, but sometimes instead after a certain number of calls (60-80), the program gets stuck and my computer starts freezing up (I guess it's still the memory problem with the plotter). :(

**Reply by akaszynski** [reply]
Just `clear()` should be sufficient.

Please enable the Python `faulthandler` and let's try to get to the root of the issue.

```py
import faulthandler
faulthandler.enable()
```

When it segfaults you'll get a useful output that I can then compare with our CI. I really appreciate your help with debugging this.

**Reply by akaszynski** [reply]
Also a `pyvista.Report()` output here would be greatly appreciated.

**Reply by akaszynski** [reply]
I’ve got it figured out (I hope) in #3245.

Please try pulling `fix/plotter_gc`

**Reply by leo2r** [reply]
Hi, sorry for being quiet for two days had a lot going on! Amazing, I'm going to test this out now 🙏🏽

**Reply by leo2r** [reply]
Oh... it seems like importing the fault handler as you suggested above actually shows the error may have to do with the boolean_intersection method and not the plotter....

```
Fatal Python error: Segmentation fault

Current thread 0x0000000113fc3600 (most recent call first):
  File "/Users/leoderiedmatten/pyvista/pyvista/core/filters/__init__.py", line 36 in _update_alg
  File "/Users/leoderiedmatten/pyvista/pyvista/core/filters/poly_data.py", line 93 in _boolean
  File "/Users/leoderiedmatten/pyvista/pyvista/core/filters/poly_data.py", line 253 in boolean_intersection
  File "/Users/leoderiedmatten/miniconda3/envs/t21-pyvista/lib/python3.7/site-packages/gym/envs/custom/pedscrewplan.py", line 122 in step
  File "train.py", line 111 in <module>
Segmentation fault: 11
```

**Reply by akaszynski** [reply]
Oh good. That's a relief (for me). Yea, booleans within VTK are flaky, I wish we had better support for booleans, and in fact there's been some work for this in https://github.com/zippy84/vtkbool.

I doubt that it will be included in VTK, and I would really like for it to be a python package on PyPI. Maybe one day.

**Reply by leo2r** [reply]
Just for the sake of completeness and so I can mark this question as answered, should I switch back to main branch and use ```pl.clear()``` and ```pv.close_all()```? And that should correctly deal with the garbage collection? Thanks :)

**Reply by leo2r** [reply]
Hi @akaszynski , coming back to you as I've now found an alternative to the boolean operation so the segfault is gone (!!), however there still seems to be some issues with the garbage collection of the Plotter because my code runs fine for the first approx 90 iterations and all of a sudden it gets super slow, and it's happening during the Plotter code... I saw you pushed a fix to the main branch less than a day ago, and I've pulled that just now :( So my question above still remains, ie: should I still use pl.clear() and pv.close_all()? Or is there another/better/working way to do it? Thanks!

**Reply by akaszynski** [reply]
> should I still use pl.clear() and pv.close_all()

That's not necessary anymore, but we do need to clean up the mappers still as we effectively create a new class when adding a mesh. I'm opening a PR for that.

**Reply by harshamarupudi56** [reply]
Hello, I seem to be having this same issue. Ran boolean operations several times and am no longer able to run my code. 

@leo2r what was the alternative you found to the boolean operations? 

https://github.com/pyvista/pyvista/issues/3290

**Reply by leo2r** [reply]
> Hello, I seem to be having this same issue. Ran boolean operations several times and am no longer able to run my code.
> 
> @leo2r what was the alternative you found to the boolean operations?
> 
> #3290

Hey, my alternative was to take the cell centres, then use the function ```select_enclosed_points()``` which returns a boolean mask (inside = 1, outside = 0) and then use that as a mask for the cell_data. Here's my code below, hopefully you can make sense of it :) Also, you can find a cleaner explanation and example at PyVista's 'Extract Cells Inside Surface' blog post! It works fine for my use case, and is significantly faster than the boolean operation :) But this was an alternative to the ```boolean_intersection()``` method, seems like you're using ```boolean_difference()``` so this solution might not work for you, sorry!

```python
bone = pv.read('L1-new.obj')
bone = bone.clean()

screw = pv.read('neww_screw_655.obj')
screw = screw.clean()
screw = screw.subdivide(1)

screw3d = screw.delaunay_3d()
screw3d_vol = screw3d.compute_cell_sizes()

screw3d_centers = screw3d.cell_centers()
screw3d_enc_points = screw3d_centers.select_enclosed_points(bone)
mask = screw3d_enc_points['SelectedPoints'].astype(bool)
inter_volume = screw3d_vol.cell_data['Volume'][mask].sum()
screw_volume = screw3d_vol.cell_data['Volume'].sum()
prop_intersect = inter_volume / screw_volume
print(prop_intersect)
```

**Reply by harshamarupudi56** [reply]
Thanks for sharing! I was intending on using intersection as well so this would be helpful to use. I'll try this out.

**Reply by leo2r** [reply]
Nice, good luck! :)

**Reply by harshamarupudi56** [reply]
Not sure what the issue is but the code seems to keep on running without returning a result. Rather then using .obj files like you did I used the pyvista library to make a cube and a sphere that I tried to intersect. 

```
import pyvista as pv 
cube = pv.Cube().triangulate().subdivide(3)
sphere = pv.Sphere(radius=0.3)

cube = cube.clean()


sphere = sphere.clean()
sphere = sphere.subdivide(1)

sphere3d = sphere.delaunay_3d()
sphere3d_vol = sphere3d.compute_cell_sizes()

sphere3d_centers = sphere3d.cell_centers()
sphere3d_enc_points = sphere3d_centers.select_enclosed_points(cube)
mask = sphere3d_enc_points['SelectedPoints'].astype(bool)
inter_volume = sphere3d_vol.cell_data['Volume'][mask].sum()
sphere_volume = sphere3d_vol.cell_data['Volume'].sum()
prop_intersect = inter_volume / sphere_volume
print(prop_intersect)
```
